### PR TITLE
LOTR - Return of The Search Filters

### PR DESCRIPTION
**Pull request by cpjmcquillan**
This PR reimplements filtering - this time in Algolia. Filtering on job_role, education phase and working patterns is possible using the new filters service. Legacy job alerts with `newly_qualified_teacher: 'true'` are also respected.

## Jira ticket URL
https://dfedigital.atlassian.net/browse/TEVA-900

## Changes in this PR
- Add algolia filters service
- Refactor algolia search into separate services
- Add facet count attribute

## Next steps
- [x]  Reindex live vacancies so job roles attribute matches array_enum

### CODE REVIEW

File: app/services/algolia/vacancy_filters_builder.rb [hunk 1]
```diff
@@ -14,6 +20,7 @@ def filter_query
     build_filters
 
     filter_array = []
+    filter_array << "(#{@search_filter})"
```
Comment (david-mears-dfe) [inline app/services/algolia/vacancy_filters_builder.rb:23]: The name `@search_filter` is maybe a little opaque in the context of these other filters. Could rename to `@published_filter`.

File: app/services/algolia/vacancy_location_builder.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
+class Algolia::VacancyLocationBuilder
```
Comment (david-mears-dfe) [inline app/services/algolia/vacancy_location_builder.rb:1]: Nice idea to break this out.

File: spec/controllers/vacancies_controller_spec.rb [hunk 3]
```diff
@@ -13,61 +13,39 @@
...
           subject
         end
       end
 
       context 'search auditor' do
         let(:params) do
-          {
-            keyword: 'Teacher'
-          }
+          { keyword: 'Teacher' }
```
Comment (cesidio) [inline spec/controllers/vacancies_controller_spec.rb:44]: I prefer one liner when is possible

```
let(:params) { { keyword: 'Teacher' } }
```

File: spec/services/algolia/vacancy_alert_builder_spec.rb [hunk 4]
```diff
@@ -10,7 +10,7 @@
   let(:default_radius) { 10 }
   let(:date_today) { Time.zone.today.to_datetime }
   let(:location_point_coordinates) { Geocoder::DEFAULT_STUB_COORDINATES }
-  let(:location_radius) { subject.convert_radius_in_miles_to_metres(default_radius) }
+  let(:location_radius) { (default_radius * 1.60934 * 1000).to_i }
```
Comment (cesidio) [inline spec/services/algolia/vacancy_alert_builder_spec.rb:13]: I have read `* 1.60934 * 1000` many times, can we define a costant for that?

Comment (cpjmcquillan) [reply spec/services/algolia/vacancy_alert_builder_spec.rb:13]: Haha yes of course

File: spec/controllers/vacancies_controller_spec.rb (hunk 3 above)
Comment (cpjmcquillan) [reply spec/controllers/vacancies_controller_spec.rb:44]: Yep me too, will make sure I change all of them 👍

**Comment by alexbowen** [comment]
in a similar vein to @cesidio comment, i see a lot of repetition of stuff like `publication_date_timestamp` and `expires_timestamp` greater than or equal to/less than this that and the other. and i wonder if this warrants some kind of filter utility or helper?

**Pull request merged**